### DISCUSSION TITLE
Apply provisional drafts in a custom controller

### DISCUSSION THREAD

**Discussion by shifuma** [craftcms/cms#15272] [Q&A] [answered]
I have a custom controller that copies the value of one field to another. I have it setup as a button in the sidebar of an entry using `EVENT_DEFINE_SIDEBAR_HTML`.

```
$myField = $entry->myField;
$entry->setFieldValue('myOtherField', $myField);
Craft::$app->elements->saveElement($entry);
```

If I make a change to the field, save the entry and then run my controller, the value is copied successfully. However, I would like to do it on click. If make a change then run trigger my controller I get _'Recent changes to the Current revision have been merged into this draft.'_ but the change isn't saved (and the field is still marked as changed.)

I thought of saving the entry once before copying my data:

```
Craft::$app->elements->saveElement($entry);
$myField = $entry->myField;
$entry->setFieldValue('myOtherField', $myField);
Craft::$app->elements->saveElement($entry);
```

But that doesn't make any difference.

I tried `Craft::$app->drafts->applyDraft($entry);` as well, but with that I get `Getting unknown property: craft\elements\Entry::draftNotes`

Is there any easy way to run my controller -> apply the provisional draft changes to the entry -> do my data copying -> resave the entry?

I hope it makes sense, but let me know if anything is unclear.

**Comment by brandonkelly** [comment]
How much work is your button/JavaScript doing? Is it copying the value directly in the form? Or just triggering an Ajax request to your controller action?

What do you want to happen if the user is already editing a provisional draft with other unsaved changes? Would you want all unsaved changes to be applied to the canonical entry?

And what if they press the button without having made any unsaved changes (so no provisional draft)? Is creating a provisional draft important, or just that the field value gets copied?

**Reply by shifuma** [reply]
Thanks for the reply. The button is just this:

```
$actionHTML .= '
  <form action="module/controller/my-action" method="post">
    <input type="hidden" name="entryId" value="' . $event->sender->canonicalId . '">
    <input type="hidden" name="CRAFT_CSRF_TOKEN" value="' . Craft::$app->getRequest()->csrfToken . '">
    <button class="btn submit" style="width: 100%">Run Action</button>
  </form>
';
```

The idea in my head is that my button is a shortcut to save an entry (and any unsaved changes that exist) then run my controller to copy the value. If there are no unsaved changes, no provisional draft is required, just copy the value.

I hope that makes sense!

Edit: I might add this is a Craft 5 site and the use case is copy values from fields inside Matrix entries -> the parent entry

**Reply by shifuma** [reply]
For now I've worked around this by using the Matrix field as an Element Index. Saving each one individually allows my controller to do it's thing without issue.

Feels like this is probably very niche so happy to close it off here.

**Comment by brandonkelly** [comment, marked as answer]
I’d just go about this by adding a `formsubmit` class to the button, which gives you an opportunity to register a new param value on the form, but otherwise let is submit the form normally.

```php
use craft\base\Element;
use craft\base\Event;
use craft\elements\Entry;
use craft\events\DefineHtmlEvent;
use craft\helpers\Html;

Event::on(Entry::class, Element::EVENT_DEFINE_SIDEBAR_HTML, function(DefineHtmlEvent $event) {
    $event->html .= Html::button('Run action', [
        'class' => ['btn', 'submit', 'formsubmit'],
        'data' => [
            'param' => 'copyValue',
            'value' => '1',
        ],
    ]);
});
```

With that in place, you could then register an after-save listener that checks if the `copyValue` param was included, and then make your changes:

```php
use craft\base\Element;
use craft\base\Event;
use craft\elements\Entry;
use craft\events\ModelEvent;
use craft\helpers\ElementHelper;

Event::on(Entry::class, Element::EVENT_AFTER_SAVE, function(ModelEvent $event) {
    /** @var Entry $entry */
    $entry = $event->sender;
    static $copyingValue = false;
    if (
        !$copyingValue &&
        !ElementHelper::isDraftOrRevision($entry) &&
        Craft::$app->request->getBodyParam('copyValue')
    ) {
        $copyingValue = true;
        $entry->targetField = $entry->sourceField;
        Craft::$app->elements->saveElement($entry);
        $copyingValue = false;
    }
});
```
